Require CC before any kind of subscription

cihangir opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Require credit card while registering teams.

## How Has This Been Tested?
Unit tests

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

Commit:
socialapi/payment: do not allow customer deletion via API

Commit:
socialapi/payment: only admins should be able to create a customer

Commit:
socialapi/payment: add CC to user before creating a sub

Commit:
socialapi/payment: unify DB operations and get the latest status of subscrption from stripe

Commit:
socialapi/payment: customer should have CC before creating a subscription

Commit:
socialapi/payment: do not return error on mongo not found cases

Commit:
socialapi/payment: charge user on their re-subscription

Commit:
server: disable team payment data check on team creation tests

cihangir commented on servers/lib/server/handlers/createteam.test.coffee:
```diff
@@ -86,9 +86,9 @@ runTests = -> describe 'server.handlers.createteam', ->
-          expect(group.payment).to.exist
-          expect(group.payment.subscription).to.exist
-          expect(group.payment.subscription.status).to.be.equal Status.TRIALING
+          # expect(group.payment).to.exist
```
cc @usirin 

lets talk about on how to change the behaviour here with your changes

rjeczalik commented on go/src/socialapi/workers/payment/api/customer.go:
```diff
@@ -11,8 +11,11 @@ import (
+	// do not allow customer deletion, causes losing track of transactions.
+	return http.StatusForbidden, nil, nil, nil
+
 	if err := checkContext(context); err != nil {
```
is it dead code or wip?

rjeczalik commented on go/src/socialapi/workers/payment/api/subscription_test.go:
```diff
@@ -21,9 +21,20 @@ func TestCreateSubscription(t *testing.T) {
-						So(subscriptionID, ShouldNotBeEmpty)
-						Convey("We should be able to get the subscription", func() {
+					withTestCreditCardToken(func(token string) {
+						updateURL := fmt.Sprintf("%s%s", endpoint, EndpointCustomerUpdate)
```
can be just `endpoint + EndpointCustomerUpdate`

rjeczalik commented on go/src/socialapi/workers/payment/api/subscription_test.go:
```diff
@@ -85,36 +96,20 @@ func TestSubscribingToPaidPlanWithWithTrialPeriodHavingNoCC(t *testing.T) {
 		withTestServer(t, func(endpoint string) {
 			withStubData(endpoint, func(username, groupName, sessionID string) {
+				withNonFreeTestPlan(func(planID string) {
+					createURL := fmt.Sprintf("%s%s", endpoint, EndpointSubscriptionCreate)
```
ditto

rjeczalik commented on go/src/socialapi/workers/payment/customer.go:
```diff
@@ -23,6 +22,8 @@ var (
 	// ErrGroupAlreadyHasSub error when a group tries to create a sub and they try to create another
 	ErrGroupAlreadyHasSub = errors.New("group already has a subscription")
+	// ErrCustomerSourceNotExists holds the error when a customer does not have a card
+	ErrCustomerSourceNotExists = errors.New("does not have source")
```
the errors message seems different from what the comment says - `errors.New("customer does not have a card"`?

cihangir replied:
in stripe. a card is a source. fixed

rjeczalik commented on go/src/socialapi/workers/payment/subscription.go:
```diff
@@ -119,30 +116,45 @@ func EnsureSubscriptionForGroup(groupName string, params *stripe.SubParams) (*st
+	}
+
+	if !hasSource {
+		return nil, ErrCustomerSourceNotExists
```
The `checkCustomerHasSource` could return `ErrCustomerSourceNotExists` if count == 0 and nil otherwise, so no need for `checkCustomerHasSource` could just have 1 return value.

cihangir replied:
i will use the difference between "having no card" & "not being able to check it" today in another feature

rjeczalik replied:
I'm not arguing against having the distinction, all I'm saying we could use error values only - the same as we use `mgo.ErrNotFound` for mongo, e.g.:

```go
err := checkCustomerHasSource(...)
if err == ErrCustomerSourceNotExists {
    // handle no-card-case
}
if err != nil {
    // handle not-able-to-check-it-case
}

cihangir replied:
oh, i misunderstood it, yeah i will change it 👍

rjeczalik commented on go/src/socialapi/workers/payment/webhook.go:
```diff
@@ -95,6 +78,17 @@ func getAmountOpts(currency string, amount int64) map[string]interface{} {
+		return ""
+	}
+
+	return fmt.Sprintf("%s%v", currencyStr, amount/100)
```
`fmt.Sprintf("%s%d", ...`

rjeczalik reviewed (approved):
👍 just minor nitpicks

Commit:
socialapi/payment: instead of formatting use string concatenation

Commit:
socialapi/paymnet: fix doc inconsistency

Commit:
socialapi/payment: instead of returning bool. return err

cihangir commented:
` Build aborted on paymentfixes ` i didnt abort the build : /

Commit:
socialapi/payment: export checkCustomerHasSource

Commit:
socialapi/payment: instead of formatting use string concatenation

Commit:
socialapi/payment: add new handler for checking if the group has a source or not

Commit:
socialapi/payment: add new location for checking if group has a credit card or not

Commit:
socialapi/payment: delete CC should delete all the sources of a group, not only the default one

Commit:
socialapi/payment: delete old CCs of a group while adding a new card

Commit:
socialapi/payment: delete old CC does not return the deleted card from on

Commit:
socialapi/payment: add test cases for CreditCardInfo

cihangir commented on go/src/socialapi/workers/payment/api/handlers.go:
```diff
@@ -14,6 +14,7 @@ const (
 	EndpointCustomerUpdate     = "/payment/customer/update"
 	EndpointCustomerDelete     = "/payment/customer/delete"
 	EndpointCreditCardDelete   = "/payment/creditcard/delete"
+	EndpointCreditCardHas      = "/payment/creditcard/has"
```
cc @usirin

usirin replied:
💯

usirin replied:
also the next line is ❤️

Commit:
workers/social: add data field

Commit:
socialapi/payment: add auth endpoint along with its tests

cihangir commented on go/src/socialapi/workers/payment/api/card.go:
```diff
@@ -49,3 +53,54 @@ func HasCreditCard(u *url.URL, h http.Header, _ interface{}, context *models.Con
 	return response.NewDefaultOK()
 }
+
+// AuthCreditCard auths some many from given source. For more info:
```
cc/ @usirin @devrim

Commit:
website: update swagger json

cihangir commented on website/swagger.json:
```diff
@@ -1388,6 +1388,9 @@
         },
         "impersonating": {
           "type": "boolean"
+        },
```
@gokmen 
```
$ scripts/wercker/run-command ./run exec scripts/api-generator.coffee --check
Swagger.json is outdated. Please run following commands to update it;

  ./run exec scripts/api-generator.coffee

and commit updated swagger.json file to current branch```

thanks for writing on how to update it 💯

Commit:
koding/db: add Data field into session

rjeczalik commented on go/src/koding/db/models/session.go:
```diff
@@ -17,4 +24,27 @@ type Session struct {
 	SessionBegan  time.Time     `bson:"sessionBegan"`
 	LastAccess    time.Time     `bson:"lastAccess"`
 	Impersonating bool          `bson:"impersonating"`
+	Data          *Data         `bson:"data,omitempty" json:"data,omitempty"`
```
It can be just `Data Data` without pointer and you could get rid of the nil check.

Commit:
socialapi/payment: remove session storage dependency

also return the charge and error to client directly, error and the response has so many valuable info for the UX

Commit:
socialapi: add testpayment into testapi

Commit:
socialapi/payment: fix test failure

Commit:
socialapi/payment: add test case for Amount, Subtotal And the Total

usirin commented:
I made an extensive test with how these endpoints are working with integrating it into the ui and all things working correctly at the moment, and since my changes in #9750 depends on this PR i am merging it.

kudos to @cihangir
